closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1124,147 +950,864 @@ private void listUsedProfile() {
+            } else if (debug != null) {
+                debug.println("Nothing to print. Value of property " + name + " is null.");
+            }
         }
+    }
 
-        /**
-         * Trim input string if not null.
-         *
-         * @param string the input string
-         * @return the string trimmed or null
-         */
-        private static String parseProperty(String string) {
-            if (string != null) {
-                string = string.trim();
-            }
+    private static final class ProfileParser {
+        // Properties specified through the profile.
+        private final Map<String, String> profileProperties;
+        private boolean descIsDefault;
+        private boolean descIsFIPS;
 
-            return string;
-        }
+        // Provider with argument (provider name + optional argument).
+        private final List<String> providers;
+        // Provider without argument.
+        private final List<String> providersSimpleName;
+        // The map is keyed by provider name.
+        private final Map<String, List<Constraint>> providerConstraints;
+
+        private final String profileID;
+
+        private Map<String, String> profilesHashes;
+        private Map<String, List<String>> profilesInfo;
+
+        private Set<String> parsedProfiles;
+
+        // The java.security properties.
+        private final Properties securityProps;
 
         /**
-         * Check if the brackets are balanced.
          *
-         * @param string input string for checking
-         * @return true if the brackets are balanced
+         * @param id    the restricted security custom profile ID
+         * @param props the java.security properties
          */
-        private static boolean areBracketsBalanced(String string) {
-            Deque<Character> deque = new LinkedList<>();
+        private ProfileParser(String id, Properties props) {
+            Objects.requireNonNull(props);
 
-            for (char ch : string.toCharArray()) {
-                switch (ch) {
-                case '{':
-                    deque.addFirst('}');
-                    break;
-                case '[':
-                    deque.addFirst(']');
-                    break;
-                case '(':
-                    deque.addFirst(')');
-                    break;
-                case '}':
-                case ']':
-                case ')':
-                    if (deque.isEmpty() || (deque.removeFirst().charValue() != ch)) {
-                        return false;
+            profileID = id;
+            securityProps = props;
+
+            profileProperties = new HashMap<>();
+
+            providers = new ArrayList<>();
+            providersSimpleName = new ArrayList<>();
+            providerConstraints = new HashMap<>();
+
+            profilesHashes = new HashMap<>();
+            profilesInfo = new HashMap<>();
+
+            hashedProfiles = new HashSet<>();
```

Comment:
I think you meant `parsedProfiles` instead of `hashedProfiles`.

Reply:
Yep, mistake. Fixed.